### PR TITLE
Uks dev

### PR DESCRIPTION
**Pull request by wxj6000**
- DF grad UHF
- DF grad UKS

### CODE REVIEW

File: gpu4pyscf/df/df_jk.py
Comment (wxj6000) [inline gpu4pyscf/df/df_jk.py:1]: @puzhichen add an example in gpu4pyscf/examples/

Comment (puzhichen) [reply gpu4pyscf/df/df_jk.py:1]: get

File: gpu4pyscf/grad/uhf.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+    from gpu4pyscf.lib.utils import to_cpu, to_gpu, device
+    
+    grad_elec = grad_elec
+    grad_nuc = rhf_grad.grad_nuc 
+    # get_veff = rhf_grad.get_veff 
+    get_veff =  get_veff
+    get_jk = rhf_grad._get_jk 
+    
+    # def get_jk(self, mol=None, dm=None, hermi=0, omega=None):
+    #     if mol is None: mol = self.mol
```
Comment (wxj6000) [inline gpu4pyscf/grad/uhf.py:340]: Delete this?

File: gpu4pyscf/grad/uks.py [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+            if ni.libxc.is_nlc(mf.xc):
+                xc = mf.xc
+            else:
+                xc = mf.nlc
+            # dma =  dm[0]
+            # dma = tag_array(dma, mo_coeff=mf.mo_coeff[0], mo_occ=mf.mo_occ[0])
+            # dmb =  dm[1]
+            # dmb = tag_array(dmb, mo_coeff=mf.mo_coeff[1], mo_occ=mf.mo_occ[1])
+            # enlc, vnlc = rks_grad.get_nlc_vxc(
+            #     ni, mol, nlcgrids, xc, dma,
```
Comment (wxj6000) [inline gpu4pyscf/grad/uks.py:93]: delete this?

File: gpu4pyscf/grad/uhf.py (hunk 1 above)
Comment (puzhichen) [reply gpu4pyscf/grad/uhf.py:340]: OK, I will clean up the codes these days.

File: gpu4pyscf/scf/uhf.py [hunk 3]
```diff
@@ -65,14 +68,8 @@ def spin_square(mo, s=1):
 
 def get_fock(mf, h1e=None, s1e=None, vhf=None, dm=None, cycle=-1, diis=None,
              diis_start_cycle=None, level_shift_factor=None, damp_factor=None):
-    if s1e is None: s1e = mf.get_ovlp()
-    if dm is None: dm = mf.make_rdm1()
-    if h1e is None: h1e = mf.get_hcore()
+    if h1e is None: h1e = cupy.asarray(mf.get_hcore())
```
Comment (wxj6000) [inline gpu4pyscf/scf/uhf.py:71]: Add those checks back.

File: gpu4pyscf/scf/uhf.py [hunk 4]
```diff
@@ -157,26 +155,54 @@ def eig(self, fock, s):
...
             vhf = vj[0] + vj[1] - vk
         else:
             ddm = cupy.asarray(dm) - cupy.asarray(dm_last)
             vj, vk = self.get_jk(mol, ddm, hermi)
             vhf = vj[0] + vj[1] - vk
-            vhf += vhf_last
+            vhf += cupy.asarray(vhf_last)
         return vhf
 
+    def scf(self, dm0=None, **kwargs):
```
Comment (wxj6000) [inline gpu4pyscf/scf/uhf.py:175]: reuse the same method in hf.py

File: gpu4pyscf/scf/uhf.py (hunk 3 above)
Comment (puzhichen) [reply gpu4pyscf/scf/uhf.py:71]: Get it.

File: gpu4pyscf/scf/uhf.py (hunk 4 above)
Comment (puzhichen) [reply gpu4pyscf/scf/uhf.py:175]: Should I make the scf in the rhf.py out of the RHF class, and reuse it in the UHF class in the uhf.py?

Comment (puzhichen) [reply gpu4pyscf/scf/uhf.py:175]: > reuse the same method in hf.py
Should I make the scf in the rhf.py out of the RHF class, and reuse it in the UHF class in the uhf.py?

**Pull request merged**